### ISSUE TITLE
Pipeline-level hooks

### ISSUE THREAD

**Issue by yuhan** [dagster-io/dagster#3613]
## Use Case
success/failure hook for a whole pipeline - single hook fired once at pipeline level, independent of a solid

## Additional Info

---
#### Message from the maintainers:
Excited about this feature? Give it a :thumbsup:. We factor engagement into prioritization.

**Comment by coderanger** [comment]
As a more concrete use case, it would be nice to have a reliable way to get notified of a failure without having to instrument every solid or scrape the GraphQL API. Success notifications are easier since that can just be a final step in the DAG, but failure is hard :)

**Comment by sryza** [comment]
Amen - especially in cases where a failure happens before any solids are executed.  E.g. if bad config is provided.

**Comment by gibsondan** [comment]
Do we have a sense of what this would look like in a possible future world where the process that does the pipeline-level orchestration is running on a system container that doesn't load user code? One possibility would be to put the logic for doing this on the Executor (or generalizing that part of the system to some unnamed 'abstraction that can reach out to workers that have access to user code')

**Comment by schrockn** [comment]
I would recommend considering plugging this into the execution plan layer. It makes a lot of sense to have pipeline failure and success hooks run as their own independent execution step. Now that our plans are more dynamic we could inject the failure hook dynamically

**Comment by coderanger** [comment]
Another option, possibly in parallel, would be to have webhooks rather than user code. So they can run more or less regardless of failures (i.e. what if the failure handler fails).

**Comment by AmirAktify** [comment]
Yes please to this one, I'd love to have a pipeline level hook.

**Comment by yuhan** [comment]
0.12.0 was released today, which comes with a first-class pipeline failure sensors. You can now write sensors to perform arbitrary actions when pipelines in your repo fail using `@pipeline_failure_sensor`. 

There are also two out-of-the-box sensors provided:
- `make_email_on_pipeline_failure_sensor` for sending emails on pipeline failure
- `make_slack_on_pipeline_failure_sensor` for sending slack messages on pipeline failure

Please check out the [Pipeline Failure Sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#pipeline-failure-sensor) docs to learn more.

cc @AmirAktify @coderanger

**Comment by coderanger** [comment]
Unfortunately using the sensor system means that can also fail as it is user provided code? Not sure why the plan of core-managed webhooks was skipped, that will make for a far more resilient system.

**Comment by yuhan** [comment]
@coderanger Thanks for the feedback. We do plan to have a core-managed mechanism to handle alerting. Will keep you posted it on that!

**Comment by coderanger** [comment]
@yuhan Is this finished somewhere else? I don't see anything merged that would solve this issue.

**Comment by yuhan** [comment]
In addition to [`@pipeline_failure_sensor`](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#pipeline-failure-sensor) mentioned above, we introduced [`@run_status_sensor`](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#run-status-sensors) recently and it provides a way to create a sensor that acts on pipeline run statuses like pipeline success, pipeline started. 

In terms of further feature requests like more resilient support, feel free to open a separate issue - we'd like to track these specific requirements separately.

So we're closing this issue as these newly introduced APIs could address most cases.
